toolchain/sem_ir/inst.h

```diff
@@ -305,19 +305,19 @@ inline auto operator<<(llvm::raw_ostream& out, TypedInst inst)
 // Associates a LocId and Inst in order to provide type-checking that the
 // TypedNodeId corresponds to the InstT.
 struct LocIdAndInst {
-  // Constructs a LocIdAndInst with no associated location. Note, we should
-  // generally do our best to associate a location for diagnostics.
+  // Constructs a LocIdAndInst with no associated location. This should be used
+  // very sparingly: only when it doesn't make sense to store a location even
+  // when the instruction kind usually has one, such as for instructions in the
+  // constants block.
   template <typename InstT>
   static auto NoLoc(InstT inst) -> LocIdAndInst {
-    return LocIdAndInst(LocId::Invalid, inst, /*is_untyped=*/true);
+    return LocIdAndInst(LocId::Invalid, inst, /*is_unchecked=*/true);
   }
 
-  // Constructs a LocIdAndInst that reuses the location associated with some
-  // other inst, typically because `inst` doesn't have an explicit
-  // representation in the parse tree.
-  template <typename InstT>
-  static auto ReusingLoc(LocId loc_id, InstT inst) -> LocIdAndInst {
-    return LocIdAndInst(loc_id, inst, /*is_untyped=*/true);
+  // Unsafely form a pair of a location and an instruction. Used in the cases
+  // where we can't statically enforce the type matches.
+  static auto UncheckedLoc(LocId loc_id, Inst inst) -> LocIdAndInst {
+    return LocIdAndInst(loc_id, inst, /*is_unchecked=*/true);
```

Comment:
Those are the only two callers right now. I can rename this to something like `UncheckedLocUnsafe` if you think this name isn't enough to discourage people from calling this function.

I want to discourage both the import cases and the non-import cases, so I don't think having an `ImportIRInstId`-specific version really helps in that sense. I could maybe do something to befriend both callers, but it's awkward because one of them is in check rather than semir.

Reply:
When you say discouraging imports, do you just mean direct calls for imports, instead pushing them through context?

Had you considered something like moving as much of the logic in context.cpp for `HasCompatibleImportedNodeKind` over here, just fetching the kind on call? i.e., yielding something like `ImportedLoc(ImportIRInstId import_ir_inst_id, InstKind imported_kind, InstT inst) -> LocIdAndInst`.

Reply:
> When you say discouraging imports, do you just mean direct calls for imports, instead pushing them through context?

Yes, exactly.

If we wanted to do the checking here, we'd also need to pass the `SemIR::File` in, and that feels like a layering issue to me -- this type seems like a lower layer than `File`. Also, the instruction and location are public members anyway, so we wouldn't be preventing them from mismatching, unless we also made other changes to encapsulate those fields (maybe we should do that, but it's a larger change than I wanted to make here).

Maybe a middle ground would be to move the function that safely builds an imported `LocIdAndInst` from `Check::Context` to `SemIR::File`.